File: TypeScript/Problem4-Palindrome.ts
function shortestPalindrome(s: string): string {
    let j = 0
    for(let i = 0 ; i < s.length; i++){
        if(s[s.length - i - 1] === s[j]){
            j++
        }
    }
    if(j==s.length)
        return s

    let p = s.substring(j);
    p = p.split('').reverse().join('');
    
    return (p + shortestPalindrome(s.substring(0, j)) + s.substring(j))
};

File: Java/Problem4-Palindrome.java
class Solution {
    public String shortestPalindrome(String s) {
        int j = 0;

        for(int i = 0; i < s.length(); i++){
            if (s.charAt(s.length() - i - 1) == s.charAt(j)){
                j++;
            }
        }

        if(j == s.length()){
            return s;
        }

        String sPart = s.substring(j);
        String sPartReversed = new StringBuilder(sPart).reverse().toString();

        return sPartReversed + shortestPalindrome(s.substring(0, j)) + sPart;
    }
}
